### ISSUE TITLE
Build error caused by BLAS (invalid conversion from void(*))

### ISSUE THREAD

**Issue by pavolmarak** [arrayfire/arrayfire#2321]
Hi everyone,

I downloaded ArrayFire sources from Github and tried to build it using CMake-GUI. I configured it to build just CPU backend. Unfortunately with no success.

This Cmake screenshot shows my CBLAS settings:
![screenshot_20181020_194854](https://user-images.githubusercontent.com/23128885/47258848-c1bcd000-d4a1-11e8-8908-04efdf7ec1d2.png)

After few seconds I get this error:

```txt
Scanning dependencies of target bin2cpp
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_intl.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_uint.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_char.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object CMakeFiles/bin2cpp.dir/CMakeModules/bin2cpp.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_uchar.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_short.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_double.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  1%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_uintl.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  2%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_ushort.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  2%] Building CXX object test/gtest/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[  2%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_int.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  2%] Building CXX object src/backend/cpu/CMakeFiles/cpu_sort_by_key_float.dir/kernel/sort_by_key/sort_by_key_impl.cpp.o
[  3%] Linking CXX executable bin2cpp
[  3%] Built target bin2cpp
[  3%] Linking CXX shared library libgtest.so
[  3%] Built target gtest
[  3%] Building CXX object test/gtest/googletest/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
[  3%] Linking CXX shared library libgtest_main.so
[  3%] Built target gtest_main
[  3%] Built target cpu_sort_by_key_double
[  3%] Built target cpu_sort_by_key_uintl
[  3%] Built target cpu_sort_by_key_int
[  3%] Built target cpu_sort_by_key_intl
[  3%] Built target cpu_sort_by_key_char
[  3%] Built target cpu_sort_by_key_ushort
[  3%] Built target cpu_sort_by_key_short
[  3%] Built target cpu_sort_by_key_uchar
[  3%] Built target cpu_sort_by_key_float
[  3%] Built target cpu_sort_by_key_uint
Scanning dependencies of target afcpu
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/canny.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/histogram.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/approx.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/fft.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/bilateral.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/iota.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/convolve.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/Array.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/anisotropic_diffusion.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/gradient.cpp.o
[  3%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/assign.cpp.o
[  4%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/blas.cpp.o
[  4%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/diff.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/diagonal.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/exampleFunction.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/homography.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/fast.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/copy.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/index.cpp.o
[  6%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/fftconvolve.cpp.o
[  6%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/hsv_rgb.cpp.o
[  6%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/harris.cpp.o
[  6%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/match_template.cpp.o
[  7%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/identity.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/cholesky.cpp.o
[  5%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/iir.cpp.o
[  6%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/join.cpp.o
[  8%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/math.cpp.o
[  8%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/inverse.cpp.o
[  8%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/lu.cpp.o
[  8%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/lookup.cpp.o
[  9%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/ireduce.cpp.o
[  9%] Building CXX object src/backend/cpu/CMakeFiles/afcpu.dir/mean.cpp.o
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp: In function ‘void (* cpu::gemm_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, cpu::scale_type<T>, cpu::cptr_type<T>, blasint, cpu::cptr_type<T>, blasint, cpu::scale_type<T>, cpu::ptr_type<T>, blasint) [with T = std::complex<float>; cpu::gemm_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const float*, const float*, int, const float*, int, const float*, float*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; cpu::scale_type<T> = const float*; cpu::cptr_type<T> = const float*; cpu::ptr_type<T> = float*]’:
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:113:10: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint) {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)}’ to ‘cpu::gemm_func_def<std::complex<float> > {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const float*, const float*, int, const float*, int, const float*, float*, int)}’ [-fpermissive]
 { return &cblas_##PREFIX##FUNC; }
          ^~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:118:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemm , cfloat  , c)
 ^~~~~~~~~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp: In function ‘void (* cpu::gemm_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, cpu::scale_type<T>, cpu::cptr_type<T>, blasint, cpu::cptr_type<T>, blasint, cpu::scale_type<T>, cpu::ptr_type<T>, blasint) [with T = std::complex<double>; cpu::gemm_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const double*, const double*, int, const double*, int, const double*, double*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; cpu::scale_type<T> = const double*; cpu::cptr_type<T> = const double*; cpu::ptr_type<T> = double*]’:
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:113:10: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint) {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)}’ to ‘cpu::gemm_func_def<std::complex<double> > {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const double*, const double*, int, const double*, int, const double*, double*, int)}’ [-fpermissive]
 { return &cblas_##PREFIX##FUNC; }
          ^~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:119:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemm , cdouble , z)
 ^~~~~~~~~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp: In function ‘void (* cpu::gemv_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, cpu::scale_type<T>, cpu::cptr_type<T>, blasint, cpu::cptr_type<T>, blasint, cpu::scale_type<T>, cpu::ptr_type<T>, blasint) [with T = std::complex<float>; cpu::gemv_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const float*, const float*, int, const float*, int, const float*, float*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; cpu::scale_type<T> = const float*; cpu::cptr_type<T> = const float*; cpu::ptr_type<T> = float*]’:
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:113:10: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint) {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)}’ to ‘cpu::gemv_func_def<std::complex<float> > {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const float*, const float*, int, const float*, int, const float*, float*, int)}’ [-fpermissive]
 { return &cblas_##PREFIX##FUNC; }
          ^~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:124:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemv , cfloat  , c)
 ^~~~~~~~~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp: In function ‘void (* cpu::gemv_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, cpu::scale_type<T>, cpu::cptr_type<T>, blasint, cpu::cptr_type<T>, blasint, cpu::scale_type<T>, cpu::ptr_type<T>, blasint) [with T = std::complex<double>; cpu::gemv_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const double*, const double*, int, const double*, int, const double*, double*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; cpu::scale_type<T> = const double*; cpu::cptr_type<T> = const double*; cpu::ptr_type<T> = double*]’:
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:113:10: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint) {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)}’ to ‘cpu::gemv_func_def<std::complex<double> > {aka void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const double*, const double*, int, const double*, int, const double*, double*, int)}’ [-fpermissive]
 { return &cblas_##PREFIX##FUNC; }
          ^~
/home/dualuser/Downloads/arrayfire/src/backend/cpu/blas.cpp:125:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemv , cdouble , z)
 ^~~~~~~~~
make[2]: *** [src/backend/cpu/CMakeFiles/afcpu.dir/build.make:128: src/backend/cpu/CMakeFiles/afcpu.dir/blas.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1185: src/backend/cpu/CMakeFiles/afcpu.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

```
My machine is Debian 9.5, GCC-6 compiler, NVidia CUDA 9.2 and OpenBLAS 0.3.3 built manually from source and installed in `/usr/local`.

This is screenshot of my `/usr/local/include`:
![screenshot_20181020_195752](https://user-images.githubusercontent.com/23128885/47258892-88d12b00-d4a2-11e8-8c4e-e61536768d37.png)

However there is also another OpenBLAS library in `/usr/lib/x86_64-linux-gnu/libopenblas.so
` installed as Debian package (version 0.3.3) and the header file  `cblas.h` is located at two more places: `/usr/include` and also `/usr/include/x86_64-linux-gnu/`.

These are BLAS-related Debian packages installed on my system:
![screenshot_20181020_200258](https://user-images.githubusercontent.com/23128885/47258947-42300080-d4a3-11e8-8b4b-3d732803408d.png)
Can you help me?
Thanks.
Paul

**Comment by pavolmarak** [comment]
Problem has been **successfully solved**.

Solution:
I have deleted all my custom OpenBLAS  installation files (libs + headers) from `/usr/local` and kept only the OpenBLAS Debian package (located in `/usr` root folder), so there is just one OpenBLAS version. I also deleted the original file `/usr/include/x86_64-linux-gnu/cblas.h`. After that I renamed the existing file `/usr/include/x86_64-linux-gnu/cblas-netlib.h` (probably from `libblas-dev` package) back to  `/usr/include/x86_64-linux-gnu/cblas.h` (now this file probably declares the correct API). 

This procedure solved my problem. Now I can build ArrayFire for all backends + Forge visualization without any problem.
